kubelet: Mark ready condition as false explicitly for terminal pods

bobbypage opened the pull request:
Terminal pods may continue to report a ready condition of true because
there is a delay in reconciling the ready condition of the containers
from the runtime with the pod status. It should be invalid for kubelet
to report a terminal phase with a true ready condition. To fix the
issue, explicitly override the ready condition to false for terminal
pods during status updates.

Signed-off-by: David Porter <david@porter.me>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #108594

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a 1.22 regression kubelet issue that could result in invalid pod status updates to be sent to the api-server where pods would be reported in a terminal phase but also report a ready condition of true in some cases.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bobbypage committed:
kubelet: Mark ready condition as false explicitly for terminal pods

Terminal pods may continue to report a ready condition of true because
there is a delay in reconciling the ready condition of the containers
from the runtime with the pod status. It should be invalid for kubelet
to report a terminal phase with a true ready condition. To fix the
issue, explicitly override the ready condition to false for terminal
pods during status updates.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/retest

robscott commented:
/retest

bobbypage commented:
See https://github.com/kubernetes/kubernetes/pull/110257 where after lowering the pollInternval in the existing regression test for https://github.com/kubernetes/kubernetes/issues/108594 we were able to detect the invalid pod status update by kubelet of (phase=Failed, Ready=True).

bobbypage commented:
/cc @smarterclayton

aojea commented:
/cc

rata commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -867,6 +867,16 @@ func mergePodStatus(oldPodStatus, newPodStatus v1.PodStatus, couldHaveRunningCon
 
+	// If the new phase is terminal, explicitly set the ready condition to false for v1.PodReady and v1.ContainersReady.
+	// It may take some time for kubelet to reconcile the ready condition, so explicitly set ready conditions to false if the phase is terminal.
+	// This is done to ensure kubelet does not report a status update with terminal pod phase and ready=true.
```
For my future self, can we link to the github issue? What do you think?

bobbypage replied:
good idea, added!

rphillips commented:
Great find!

rphillips commented:
/kind bug
/triage accepted
/priority important-soon

rphillips commented:
@bobbypage I'm testing this patch and terminating a pod. I still seeing the pod in ready status during the termination phase.

```
apiVersion: v1
kind: Pod
metadata:
  name: bash
spec:
  containers:
  - name: bash
    image: bash
    command:
    - bash
    - -c
    - "_term() { rm -f /tmp/ready ; sleep 30; exit 0; } ; trap _term SIGTERM; touch /tmp/ready ; while true; do echo \"hello\"; sleep 10; done"
  terminationGracePeriodSeconds: 30
  restartPolicy: Never
```

bobbypage commented:
> @bobbypage I'm testing this patch and terminating a pod. I still seeing the pod in ready status during the termination phase.
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: bash
> spec:
>   containers:
>   - name: bash
>     image: bash
>     command:
>     - bash
>     - -c
>     - "_term() { rm -f /tmp/ready ; sleep 30; exit 0; } ; trap _term SIGTERM; touch /tmp/ready ; while true; do echo \"hello\"; sleep 10; done"
>   terminationGracePeriodSeconds: 30
>   restartPolicy: Never
> ```

That should be expected. The PR doesn't change the semantics around reporting `ready=false` while a pod is terminating itself. The PR does *not* address https://github.com/kubernetes/kubernetes/issues/102537#issue-909658949 which is a separate issue and as discussed it's unclear if we want to change the behavior for that (xref: https://github.com/kubernetes/kubernetes/pull/110191#issuecomment-1142294392).

 The main change in behavior for the PR is that after this patch, kubelet will never report ready=true and a terminal phase (i.e. phase=Succeeded/Failed) during a status update. For example, the status update caught in the regression test here (https://github.com/kubernetes/kubernetes/pull/110257#issuecomment-1140115944) is something we want to prevent and ensure kubelet never reports such as invalids state to the api server. The general invariant is that ready=true should only be reported for non-terminal phases.

bobbypage commented on pkg/kubelet/status/generate.go:
```diff
@@ -69,6 +71,15 @@ func GenerateContainersReadyCondition(spec *v1.PodSpec, containerStatuses []v1.C
 		}
 	}
 
+	// If the pod phase is failed, explicitly set the ready condition to false for containers since they may be in progress of terminating.
```
It does and a previous PR ensured we only report terminal phase after containers are terminated. However, the status of the containers come from the runtime which can be out of date and stale, so kubelet should explicitly set  the readiness to false if the pod will be reported as terminal.

249043822 commented:
/cc
From my testing, this change look good to me.

bobbypage committed:
test: update graceful node shutdown e2e with watch

Use a watch to detect invalid pod status updates in graceful node
shutdown node e2e test. By using a watch, all pod updates will be
captured while the previous logic required polling the api-server which
could miss some intermediate updates.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/test pull-kubernetes-node-kubelet-serial-containerd

bobbypage commented:
Pushed updated revision that includes update to node e2e test which includes a reproducer for #108594. Same test on master currently fails (see https://github.com/kubernetes/kubernetes/pull/110413) while the serial job is passing with this PR.

bobbypage commented:
/assign @rphillips 

Ryan, can you PTAL again? Thanks!

rphillips commented:
/lgtm

rphillips commented:
/assign @mrunalp

rphillips commented:
@bobbypage release note?

bobbypage commented:
added release note, thanks!

dchen1107 commented:
/approve based on @rphillips and @mrunalp reviews.

thockin commented:
xref https://github.com/kubernetes/kubernetes/pull/110255

This feels like a bandaid.  Consumers which do not handle terminal `Phase`s might see these Pods as "alive but not serving at the moment", when in fact they are gone and never coming back.

Consider what might have happened if this was done in the first place - we'd have EndpointSlices with potentially tons of "unready" endpoints (we track both ready and not).  It would not have been the same problem we actually had, but it would have been a slow-burning "leak".

It seems like the REAL fix is to a) document the pod lifecycle contract; and b) fix _all_ consumers.

That said, I guess we've had this problem FOREVER, right?  If you are examining pods and not handling Phase properly, this doesn't make you worse than you were before.

thockin commented on pkg/kubelet/status/generate.go:
```diff
@@ -69,6 +71,15 @@ func GenerateContainersReadyCondition(spec *v1.PodSpec, containerStatuses []v1.C
 	}
 
+	// If the pod phase is failed, explicitly set the ready condition to false for containers since they may be in progress of terminating.
+	if podPhase == v1.PodFailed {
```
Why is this not also checking `PodSucceeded` ?  Should we leave a comment?

thockin replied:
And why does this not call `generateContainersReadyConditionForTerminalPhase(podPhase)` ?

bobbypage replied:
> Why is this not also checking PodSucceeded ? Should we leave a comment?

PodSucceded is already handled above on line 66, see [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/generate.go#L64-L70)

> And why does this not call generateContainersReadyConditionForTerminalPhase(podPhase) ?

I created that helper function to be used in `status_manager.go` to be able to handle either phase. I returned the condition directly inline here, since it's what done in the succeeded case above, but I agree with you, we can switch it to use the helper function as well.

bobbypage replied:
I updated, based on your feedback and reused `generateContainersReadyConditionForTerminalPhase` for both succeeded and failed cases, so the condition generation logic is in one place.

bobbypage commented:
Thanks Tim! I agree with you, ideally clients should ignore terminal pods, and if the pod is terminal ignore the ready condition in the first place. We can't guarantee that all clients will do that though.

However, at least pre-1.22 (before this regression was introduced), clients could assume that terminal pods would always report ready=false, which is no longer the case. This PR aims to revert to that behavior and ensure kubelet never reports an "invalid" status update of `Phase=Terminal`, `Ready=True`. 

> That said, I guess we've had this problem FOREVER, right? If you are examining pods and not handling Phase properly, this doesn't make you worse than you were before.

Yup, this doesn't make it any worse for clients ignoring phases but rather brings back the guarantee that terminal pods will always report a ready status of false.

thockin reviewed (commented):
/approve

I'm a big fan of leaving future-me clues.

We should move directories around so that podutil OWNERs don't need API approvers.

bobbypage commented:
/hold

to update based on feedback

bobbypage commented:
/unhold

Pushed an additional revision to fix suggestions [here](https://github.com/kubernetes/kubernetes/pull/110256#discussion_r892941368)

thockin commented:
Assuming previous LGTM holds after minor changes:

/lgtm

bobbypage commented:
/retest

bobbypage commented:
Cherrypicks:

1.24 - https://github.com/kubernetes/kubernetes/pull/110479
1.23 - https://github.com/kubernetes/kubernetes/pull/110480
1.22 - https://github.com/kubernetes/kubernetes/pull/110481

smarterclayton commented on pkg/kubelet/status/generate.go:
```diff
@@ -62,11 +64,12 @@ func GenerateContainersReadyCondition(spec *v1.PodSpec, containerStatuses []v1.C
-			Status: v1.ConditionFalse,
-			Reason: PodCompleted,
-		}
+		return generateContainersReadyConditionForTerminalPhase(podPhase)
```
Going back and looking through this, the `len(unknownContainers) == 0` above has me worried.

There are two phases in play - the apiserver phase, and the one the Kubelet tracks.  Setting phase to terminal on the apiserver is roughly the same outcome as deleting the pod - the kubelet should inexorably converge the actual pod state to stopped.  Likewise if the Kubelet observes a failure or evicts the pod, the phase should inexorably converge.  However, internal to the Kubelet status loop the phase should *match* the actual observed state of the pod containers, and so we should never be in a terminal state here without these containers being in some state we recognize (I think).

This is something I want to have laid out in the pod lifecycle clarification KEP so we can look at all the factors and eliminate my "shoulds" from above - it might be that the correct status when succeeded + > 1 unknown container = pod ready = false here, always.  But it's super subtle, and we need this sort of stuff in a clear top level doc, not in comment threads in a PR.   I'll add it to the list.
